lib/ret_web/controllers/api/v1/project_assets_controller.ex

```diff
@@ -0,0 +1,62 @@
+defmodule RetWeb.Api.V1.ProjectAssetsController do
+  use RetWeb, :controller
+  import Ecto.Query
+
+  alias Ecto.{Multi}
+  alias Ret.{Asset, Project, ProjectAsset, Repo, Storage}
+
+  # Limit to 1 TPS
+  plug(RetWeb.Plugs.RateLimit when action in [:create])
+
+  def index(conn, %{"id" => project_sid}) do
+    account = Guardian.Plug.current_resource(conn)
+
+    case get_project(account, project_sid) do
+      %Project{} = project -> render(conn, "index.json", assets: project.assets)
+      nil -> conn |> send_resp(404, "Project not found")
+    end
+  end
+
+  def create(conn, %{"id" => project_sid, "asset" => params}) do
+    account = conn |> Guardian.Plug.current_resource()
+
+    case get_project(account, project_sid) do
+      %Project{} = project -> create(conn, params, account, project)
+      nil -> conn |> send_resp(404, "Project not found")
+    end
+  end
+
+  defp create(conn, params, account, project) do
+    case Storage.promote(params["file_id"], params["access_token"], nil, account) do
+      {:ok, asset_owned_file} ->
+
+        asset_changeset = Asset.changeset(%Asset{}, account, asset_owned_file, params)
```

Comment:
i'd move this operation into a function on `Ret.Asset` that returns the result of `Repo.transaction`